typing/typecore.ml

```diff
@@ -569,7 +569,14 @@ let enter_orpat_variables loc env  p1_vs p2_vs =
   unify_vars p1_vs p2_vs
 
 let rec build_as_type (env : Env.t) p =
-  let as_ty = build_as_type_aux env p in
+  let has_ground_constraint =
+    List.exists
+      (function Tpat_constraint cty, _, _ -> free_variables cty.ctyp_type = []
+        | _ -> false)
+      p.pat_extra
+  in
+  let as_ty =
+    if has_ground_constraint then newgenvar () else build_as_type_aux env p in
```

Comment:
Two remarks:

1. There should be a comment here that points to the current PR, otherwise people reading the code in the future will have a hard time guessing why we special-case on ground constraints.

2. It is not clear when reading the code why `newgenvar()` is a reasonable choice here. I would have rather expected (an instance of) the `cty.ctyp_type` that was found above. If I understand correctly, the two are in fact equivalent because of the unification that happens below. This makes the code a bit convoluted in my opinion. I think you should either:
  - rewrite it to be clearer
  - or add a comment to explain why `newgenvar()` is acceptable here

Here would be a potential suggestion for a clearer implementation (sketch, untested).

```ocaml
  (* 11799, 12313: in the case of [(p : ty) as x] where [ty] is a ground type,
     we can return [ty] right away. *)
  let ground_constraint =
    List.find_map (function
    | Tpat_constrat {ctyp_type = ty}; _, _, _ when free_variables ty = [] -> Some ty
    | _ -> None
    ) p.pat_extra
  in
  match ground_constraint with
  | Some ty -> instance ty
  | None ->
     let as_ty = build_as_type_aux env p in
     List.fold_left ...
```

Reply:
(Actually pointing at https://github.com/ocaml/ocaml/pull/11799#issuecomment-1597363236 could be even more informative)

Reply:
The code you suggest is actually wrong: there might also be non-ground type annotations, and they might result in a different type from the point of view of ambivalence, so we must keep them. So better explain that do some extra work.
I added comments, in particular for the clearly wrong statement about unification below.

Reply:
Okay, new attempt:

```ocaml
let rec build_as_type (env : Env.t) p =
  build_as_type_extra env p p.pat_extra

and build_as_type_extra env p = function
  | [] -> build_as_type_aux env p
  | (Tpat_type _ | Tpat_open _ | Tpat_unpack) :: rest ->
    build_as_type_extra env p rest
  | Tpat_constraint {cty_type = ty; _} :: _ when free_variables ty = [] ->
    (* TODO some explanation for this special case here *)
    instance ty
  | Tpat_constraint cty :: rest ->
    let as_ty = build_as_type_extra env p rest in
    (* [generic_instance] can only be used if the variables of the original
       type ([cty.ctyp_type] here) are not at [generic_level], which they are
       here.
       If we used [generic_instance] we would lose the sharing between
       [instance ty] and [ty].  *)
    let ty =
      with_local_level ~post:generalize_structure
        (fun () -> instance cty.ctyp_type)
    in
    (* This call to unify may only fail due to missing GADT equations *)
    unify_pat_types p.pat_loc env (instance as_ty) (instance ty);
    ty
```

Reply:
OK, it goes against my policy of minimal textual change, but your code is fine so I'll use it...
Feels a bit CPS though..

Reply:
Followed suggestion and rebased.